rkt/attach.go

```diff
@@ -0,0 +1,181 @@
+// Copyright 2016 The rkt Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/coreos/rkt/common"
+	pkgPod "github.com/coreos/rkt/pkg/pod"
+	"github.com/coreos/rkt/stage0"
+	"github.com/coreos/rkt/store/imagestore"
+	"github.com/coreos/rkt/store/treestore"
+	"github.com/spf13/cobra"
+)
+
+var (
+	cmdAttach = &cobra.Command{
+		Use:   "attach [--app=APPNAME] [--mode=MODE] UUID",
+		Short: "Attach to an app running within a rkt pod",
+
+		Long: `UUID should be the UUID of a running pod.`,
+		Run:  ensureSuperuser(runWrapper(runAttach)),
+	}
+	flagAttachMode string
+)
+
+func init() {
+	if common.IsExperimentEnabled("attach") {
+		cmdRkt.AddCommand(cmdAttach)
+		cmdAttach.Flags().StringVar(&flagAppName, "app", "", "name of the app to enter within the specified pod")
+		cmdAttach.Flags().StringVar(&flagAttachMode, "mode", "list", "attach mode")
+	}
+}
+
+func runAttach(cmd *cobra.Command, args []string) (exit int) {
+	if len(args) < 1 {
+		cmd.Usage()
+		return 254
+	}
+
+	uuid := args[0]
+	p, err := pkgPod.PodFromUUIDString(getDataDir(), uuid)
+	if err != nil {
+		stderr.PrintE("problem retrieving pod", err)
+		return 254
+	}
+	defer p.Close()
+
+	if p.State() != pkgPod.Running {
+		stderr.Printf("pod %q isn't currently running", p.UUID)
```

Comment:
👍 for bubbling up so defers happen; I don't think it needs or should be an error object either necessarily since one step up won't have more info in this case ever.